libs/common/src/api/contract.ts

```diff
@@ -1,73 +1,55 @@
 import { initContract } from "@ts-rest/core";
 import { z } from "zod";
 import { extendZodWithOpenApi } from "@anatine/zod-openapi";
+import * as Rx from "../rxdefs.js";
+import RxUtils from "../rxutils.js";
+import { Iso639Set1Codes } from "../iso639-1.js";
 
 extendZodWithOpenApi(z);
 
 const c = initContract();
 
-/** A regex testing for an *URL-safe* base64 string (RFC4648 sect 5) */
-const UrlSafeBase64Rx = /^[A-Z0-9_-]+$/i;
-
-/** A string regular expression matching an XML NCName
- *
- * - An NCName is an XML Name, but with no colons allowed.
- * - An XML Name is a complicated beast, but loosely a Unicode version of \w,
- *   or in other words, a unicode alphanumeric symbolic identifier.
- *   - It can contain digits, letters, hyphens, periods and underscores and
- *     certain unicode equivalents.
- *   - But it must not start with digits, a hyphen or a period, nor certain
- *     unicode equivalents.
- *
- * The regex for an XML Name is adapted from O'Reilly Regex Cookbook section 8.4,
- * "XML 1.0 names (approximate)" - but the colon is removed.
- *
- * Paraphrasing that book's explanation:
- * - the name start character can be a [:_] or
- * - any of the following Unicode categories:
- *   - Lowercase letter (Ll)
- *   - Uppercase letter (Lu)
- *   - Titlecase letter (Lt)
- *   - Letter without case (Lo)
- *   - Letter number (Nl)
- * - subsequent characters can also include [.-] or
- *   - Mark (M)
- *   - Modifier letter (Lm)
- *   - Decimal digit (Nd)
+/** Helper function to generate Zod refinements from a RegExp
  *
- * This definition is not compiled, as it is intended for composition below.
- * Therefore it is wrapped in a non-capturing group to isolate it without affecting
- * the captures which may be defined around it.
+ * It promotes the RegExp to be a unicode, entire-string match, if it is
+ * not already.
  *
- * This regex requires node 10+ to be able to use /u and \p.
+ * It also sets the validation error message attribute from the message parameter.
  *
+ * @returns a Zod validator generated by the Zod.string().refine() method
  */
-const NCName =
-  "(?:[_\\p{Ll}\\p{Lu}\\p{Lt}\\p{Lo}\\p{Nl}][_.\\p{L}\\p{M}\\p{Nd}\\p{Nl}-]*)";
-
-/** Match a QName
- *
- * Paraphrasing https://en.wikipedia.org/wiki/QName
- * - A QName is an NCName (see above)
- * - Or two of them delimited by a colon.
- *
- * Note: for our purposes, we *require* a colon delimiter - it can't just be a NCName.
- * This is because we need an abbreviation with which to look up the URL prefix.
- *
- * FIXME Perhaps we should also disallow common URI scheme prefixes.
- * FIXME Maybe we don't care about unicode?
- *
- * This regex requires node 10+ to be able to use /u and \p.
- *
- */
-const QNameRx = new RegExp(`^${NCName}[:]${NCName}$`, "gsu");
+function ZodRegex(rx: RegExp, message: string) {
+  return z.string().refine((v: any) => RxUtils.uaon(rx).test(String(v)), {
```

Comment:
@wu-lee What's the reason for using `.refine` vs `.regex` since the former looks more complicated and verbose? This is a question rather than a piece of feedback, I'm new to Zod

Reply:
Hmm I think I might understand this... it's so that the regex can be built up programatically rather than statically hardcoded as regex string?

I feel like pre-building outside of the code then hardcoding the regex, and simply commenting what it means, would be less confusing for readers. But I suppose it makes it harder to tweak the regex later, since the work of building it would need to be redone from scratch, so this way makes sense

Reply:
> @wu-lee What's the reason for using `.refine` vs `.regex` since the former looks more complicated and verbose? This is a question rather than a piece of feedback, I'm new to Zod

I was using `.regex()` but adopted `.refine()` from somewhere I found researching Zod - I forget where, perhaps an issue or the documentation (which does have usage examples I think I used).

The main advantage I can see here is that the error message on failure states the reason more clearly - i.e. I think you get "This is not a valid QName" vs "This does not match the regular expression <insert noise here>".

Another advantage here is that it does some wrapping of the pattern: it adds anchors to it. This is so that the regular expressions defined in the code can all be kept unanchored patterns. Which means they can be re-used as components. For instance, Rx.NCName is used to build Rx.QName, which it couldn't be if it were anchored to match the whole string. Likewise Rx.QName is used in Rx.PrefixURI's definition.

Reply:
> Hmm I think I might understand this... it's so that the regex can be built up programatically rather than statically hardcoded as regex string?

Yes. And as I say, the bits are then reusable, and NCName can be used to define QName, which can be used to define PrefixURI.
 
> I feel like pre-building outside of the code then hardcoding the regex, and simply commenting what it means, would be less confusing for readers. But I suppose it makes it harder to tweak the regex later, since the work of building it would need to be redone from scratch, so this way makes sense

I started doing that, then realised how much duplication there was. The regexes got unwieldy quickly.

Reply:
Thanks, this makes sense